File: backend/app/core/hybrid_crypto.py
"""Hybrid (asymmetric) envelope/bundle encryption — protocol v3 (ECIES).

Why this exists
---------------
The legacy envelope/bundle crypto derives its AES key from a SHARED password
(SYNC_PASSWORD) that must live on BOTH machines. On the work PC that shared
secret is exposed to anyone who can read the process/config, and once it leaks
*all* recorded traffic (past and future) is decryptable. There is no forward
secrecy.

v3 replaces the shared secret on the *upload* path with public-key crypto:

  * The home server owns a long-term X25519 static keypair. Only its PUBLIC
    key is distributed to the work PC (pinned). A public key is not a secret.
  * For every API call the work PC generates a throwaway (ephemeral) X25519
    keypair, does ECDH against the pinned server public key, and derives the
    session keys from the shared secret. The ephemeral private key is discarded
    after the call.

Consequences for the threat model (adversary on the work PC):
  * Reading everything on the work PC later (config, disk image, even RAM
    *between* syncs) yields only the server PUBLIC key + discarded ephemeral
    publics — useless to decrypt any recorded upload. This is the headline win.
  * Forward secrecy holds with respect to the work PC: each call uses a fresh
    ephemeral that is gone afterwards.
  * NOTE: this is ECIES (ephemeral-static). It is *not* forward-secret against
    compromise of the home server's STATIC private key — an attacker holding
    that key plus recorded traffic can recompute the shared secret (the
    ephemeral public is on the wire). The home PC is the trusted recipient, so
    that is an accepted tradeoff. Full FS against the recipient key would need
    an interactive ephemeral-ephemeral handshake (extra round trip).
  * Client authentication is still provided by the API key (Bearer). Leaking
    the API key only lets an attacker WRITE to the server; it does not break
    confidentiality of recorded traffic.
  * Unbreakable wall (unchanged): plaintext + an active session key must
    coexist in RAM *during* a sync. A memory dump taken mid-sync defeats any
    scheme. v3 does not claim to fix that.

Primitives (all available in both Python `cryptography` and JDK 17 JCA):
  * X25519 ECDH
  * HKDF-SHA256  (salt = ephemeral public key, info = direction/purpose label)
  * AES-256-GCM  (fresh 12-byte nonce per message)

Wire formats
------------
Envelope (the "e" string), base64:  nonce[12] || AES-GCM(ct+tag)
Bundle (raw bytes / base64 "d"):     nonce[12] || AES-GCM(ct+tag)
The ephemeral public key is transported *separately* (JSON field "epk" or
multipart form field "k"), NOT prefixed onto the ciphertext. Presence of the
ephemeral public key is what selects v3 over the legacy password path, so the
discriminator is unambiguous (legacy requests never carry "epk"/"k").

HKDF info labels (must match the Kotlin client byte-for-byte):
  request  envelope (client -> server):  b"lgm/v3/env/req"
  response envelope (server -> client):  b"lgm/v3/env/resp"
  request  bundle   (client -> server):  b"lgm/v3/bundle/req"
  response bundle   (server -> client):  b"lgm/v3/bundle/resp"
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
import stat
from pathlib import Path

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

_NONCE_SIZE = 12
_KEY_SIZE = 32
_X25519_PUB_SIZE = 32

# Direction/purpose labels for HKDF — keep in sync with HybridCrypto.kt.
INFO_ENV_REQ = b"lgm/v3/env/req"
INFO_ENV_RESP = b"lgm/v3/env/resp"
INFO_BUNDLE_REQ = b"lgm/v3/bundle/req"
INFO_BUNDLE_RESP = b"lgm/v3/bundle/resp"


# ─────────────────────────── server key management ──────────────────────────


def load_or_create_server_key(path: Path) -> X25519PrivateKey:
    """Load the server's long-term X25519 private key, creating it on first use.

    The key is stored as 32 raw bytes with 0600 permissions. Returns the
    private key object.
    """
    path = Path(path)
    if path.exists():
        raw = path.read_bytes()
        if len(raw) != 32:
            raise ValueError(f"Corrupt server key at {path}: expected 32 bytes, got {len(raw)}")
        return X25519PrivateKey.from_private_bytes(raw)

    priv = X25519PrivateKey.generate()
    raw = priv.private_bytes_raw()
    path.parent.mkdir(parents=True, exist_ok=True)
    # Write then tighten perms (no-op on Windows but harmless).
    path.write_bytes(raw)
    try:
        os.chmod(path, stat.S_IRUSR | stat.S_IWUSR)
    except Exception:
        pass
    return priv


def public_bytes(priv: X25519PrivateKey) -> bytes:
    """Raw 32-byte X25519 public key."""
    return priv.public_key().public_bytes_raw()


def public_b64(priv: X25519PrivateKey) -> str:
    """URL-safe base64 of the raw public key (no padding stripped — standard)."""
    return base64.urlsafe_b64encode(public_bytes(priv)).decode("ascii")


def fingerprint(pub: bytes) -> str:
    """Short human-comparable fingerprint: first 8 bytes of SHA-256, hex, grouped."""
    digest = hashlib.sha256(pub).hexdigest()[:16]
    return " ".join(digest[i : i + 4] for i in range(0, 16, 4))


# ─────────────────────────────── core ECIES ─────────────────────────────────


def _derive(shared: bytes, epk: bytes, info: bytes) -> bytes:
    """HKDF-SHA256 → 32-byte AES key, salted with the ephemeral public key."""
    return HKDF(
        algorithm=hashes.SHA256(),
        length=_KEY_SIZE,
        salt=epk,
        info=info,
    ).derive(shared)


def _seal(key: bytes, plaintext: bytes) -> bytes:
    nonce = os.urandom(_NONCE_SIZE)
    ct = AESGCM(key).encrypt(nonce, plaintext, None)
    return nonce + ct


def _open(key: bytes, blob: bytes) -> bytes:
    if len(blob) < _NONCE_SIZE + 16:
        raise ValueError("ciphertext too short")
    nonce, ct = blob[:_NONCE_SIZE], blob[_NONCE_SIZE:]
    return AESGCM(key).decrypt(nonce, ct, None)


class HybridServerContext:
    """Server-side per-request crypto context bound to one client ephemeral key.

    The same shared secret encrypts the response, so a single context handles
    both directions of one API call.
    """

    def __init__(self, server_priv: X25519PrivateKey, epk: bytes):
        if len(epk) != _X25519_PUB_SIZE:
            raise ValueError("invalid ephemeral public key length")
        self.epk = epk
        self._shared = server_priv.exchange(X25519PublicKey.from_public_bytes(epk))

    # — metadata envelope —
    def open_envelope(self, b64: str) -> dict:
        key = _derive(self._shared, self.epk, INFO_ENV_REQ)
        plaintext = _open(key, base64.b64decode(b64))
        return json.loads(plaintext.decode("utf-8"))

    def seal_envelope(self, payload: dict) -> str:
        key = _derive(self._shared, self.epk, INFO_ENV_RESP)
        plaintext = json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
        return base64.b64encode(_seal(key, plaintext)).decode("ascii")

    # — git bundle bytes —
    def open_bundle(self, blob: bytes) -> bytes:
        key = _derive(self._shared, self.epk, INFO_BUNDLE_REQ)
        return _open(key, blob)

    def seal_bundle(self, plaintext: bytes) -> bytes:
        key = _derive(self._shared, self.epk, INFO_BUNDLE_RESP)
        return _seal(key, plaintext)


def decode_epk(epk_b64: str) -> bytes:
    """Decode a base64 (standard or url-safe) ephemeral public key to 32 bytes."""
    s = epk_b64.strip()
    # Accept both alphabets; add padding if the client stripped it.
    s = s.replace("-", "+").replace("_", "/")
    s += "=" * (-len(s) % 4)
    raw = base64.b64decode(s)
    if len(raw) != _X25519_PUB_SIZE:
        raise ValueError("invalid ephemeral public key length")
    return raw

File: idea-plugin/src/main/kotlin/localgitmirror/idea/workkit/HybridCrypto.kt
package localgitmirror.idea.workkit

import java.math.BigInteger
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.SecureRandom
import java.security.interfaces.XECPublicKey
import java.security.spec.NamedParameterSpec
import java.security.spec.XECPublicKeySpec
import java.util.Base64 as JavaBase64
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Hybrid (asymmetric) envelope/bundle crypto — protocol v3 (ECIES).
 *
 * Replaces the shared SYNC_PASSWORD on the UPLOAD path with public-key crypto:
 *   * The home server owns a long-term X25519 static keypair; only its PUBLIC
 *     key is pinned on this (work) machine. A public key is not a secret.
 *   * Each API call uses a fresh throwaway (ephemeral) X25519 keypair, does
 *     ECDH against the pinned server public key, derives the session keys, then
 *     discards the ephemeral private key.
 *
 * Why this matters for an adversary on THIS machine:
 *   * Reading everything here later (config / disk / RAM between syncs) yields
 *     only the server PUBLIC key + spent ephemeral publics — useless to decrypt
 *     any recorded upload. Forward secrecy holds w.r.t. this machine.
 *   * The only thing that can decrypt an upload is the server's STATIC private
 *     key, which never leaves the home PC.
 *   * NOT defended: a memory dump taken DURING an active sync (plaintext + a
 *     live session key must coexist in RAM then — true of any scheme).
 *
 * MUST stay byte-for-byte compatible with backend/app/core/hybrid_crypto.py:
 *   * X25519 ECDH → 32-byte raw shared secret
 *   * HKDF-SHA256(ikm = shared, salt = ephemeral_pub[32], info = label) → 32B
 *   * AES-256-GCM, 12-byte random nonce, wire = nonce[12] || ct(+tag)
 *
 * HKDF info labels (identical to the Python side):
 *   request  envelope (client -> server):  "lgm/v3/env/req"
 *   response envelope (server -> client):  "lgm/v3/env/resp"
 *   request  bundle   (client -> server):  "lgm/v3/bundle/req"
 *   response bundle   (server -> client):  "lgm/v3/bundle/resp"
 */
object HybridCrypto {

  private const val NONCE_SIZE = 12
  private const val GCM_TAG_BITS = 128
  private const val KEY_SIZE = 32
  private const val PUB_SIZE = 32

  private val INFO_ENV_REQ = "lgm/v3/env/req".toByteArray(Charsets.US_ASCII)
  private val INFO_ENV_RESP = "lgm/v3/env/resp".toByteArray(Charsets.US_ASCII)
  private val INFO_BUNDLE_REQ = "lgm/v3/bundle/req".toByteArray(Charsets.US_ASCII)
  private val INFO_BUNDLE_RESP = "lgm/v3/bundle/resp".toByteArray(Charsets.US_ASCII)

  /** Decode a pinned server public key from base64 (standard or url-safe). */
  fun decodeServerPub(b64: String): ByteArray {
    val raw = decodeBase64Flexible(b64)
    require(raw.size == PUB_SIZE) { "Invalid server public key length: ${raw.size}" }
    return raw
  }

  /**
   * One ephemeral key bound to one API call. Seals the request and opens the
   * response with the SAME shared secret (different HKDF labels per direction).
   * Create once per call; the ephemeral private key lives only inside this
   * object and is dropped when it is garbage-collected. Call [wipe] when done.
   */
  class Session private constructor(
    val ephemeralPub: ByteArray,
    private val shared: ByteArray,
  ) {
    /** Ephemeral public key, url-safe base64 (no padding) — sent as "epk"/"k". */
    val epkB64: String get() = JavaBase64.getUrlEncoder().withoutPadding().encodeToString(ephemeralPub)

    fun sealEnvelope(jsonStr: String): String {
      val key = hkdf(shared, ephemeralPub, INFO_ENV_REQ)
      return JavaBase64.getEncoder().encodeToString(seal(key, jsonStr.toByteArray(Charsets.UTF_8)))
    }

    fun openEnvelope(b64: String): String {
      val key = hkdf(shared, ephemeralPub, INFO_ENV_RESP)
      return String(open(key, JavaBase64.getDecoder().decode(b64)), Charsets.UTF_8)
    }

    /** Seal raw bundle bytes for upload (the multipart attachment). */
    fun sealBundle(data: ByteArray): ByteArray {
      val key = hkdf(shared, ephemeralPub, INFO_BUNDLE_REQ)
      return seal(key, data)
    }

    /** Open a bundle returned by the server (export "d"). */
    fun openBundle(blob: ByteArray): ByteArray {
      val key = hkdf(shared, ephemeralPub, INFO_BUNDLE_RESP)
      return open(key, blob)
    }

    /** Best-effort: zero the shared secret. (JVM may keep copies; not guaranteed.) */
    fun wipe() {
      java.util.Arrays.fill(shared, 0)
    }

    companion object {
      fun create(serverPub: ByteArray): Session {
        require(serverPub.size == PUB_SIZE) { "Invalid server public key length" }
        val kpg = KeyPairGenerator.getInstance("XDH")
        kpg.initialize(NamedParameterSpec.X25519)
        val kp = kpg.generateKeyPair()

        val ka = KeyAgreement.getInstance("XDH")
        ka.init(kp.private)
        ka.doPhase(rawToPublicKey(serverPub), true)
        val shared = ka.generateSecret()

        val epk = publicKeyToRaw(kp.public as XECPublicKey)
        return Session(epk, shared)
      }
    }
  }

  // ───────────────────────────── primitives ─────────────────────────────────

  private fun seal(key: ByteArray, plaintext: ByteArray): ByteArray {
    val nonce = ByteArray(NONCE_SIZE).also { SecureRandom().nextBytes(it) }
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(GCM_TAG_BITS, nonce))
    return nonce + cipher.doFinal(plaintext)
  }

  private fun open(key: ByteArray, blob: ByteArray): ByteArray {
    require(blob.size >= NONCE_SIZE + 16) { "ciphertext too short" }
    val nonce = blob.copyOfRange(0, NONCE_SIZE)
    val ct = blob.copyOfRange(NONCE_SIZE, blob.size)
    val cipher = Cipher.getInstance("AES/GCM/NoPadding")
    cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(GCM_TAG_BITS, nonce))
    return cipher.doFinal(ct)
  }

  /**
   * HKDF-SHA256 → 32-byte key. salt = ephemeral public key, info = label.
   * One expand block is enough for a 32-byte output. Matches Python's
   * cryptography.hazmat HKDF(salt=epk, info=label, length=32).
   */
  internal fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray): ByteArray {
    val mac = Mac.getInstance("HmacSHA256")
    // Extract: PRK = HMAC(salt, ikm)
    mac.init(SecretKeySpec(salt, "HmacSHA256"))
    val prk = mac.doFinal(ikm)
    // Expand: T(1) = HMAC(PRK, info || 0x01); OKM = T(1)[0..32)
    mac.init(SecretKeySpec(prk, "HmacSHA256"))
    mac.update(info)
    mac.update(0x01.toByte())
    val t1 = mac.doFinal()
    return t1.copyOf(KEY_SIZE)
  }

  // ───────────────── X25519 raw <-> JCA key conversions ─────────────────────
  //
  // Stock JDK 17 has no raw X25519 byte API, so we convert through the
  // u-coordinate BigInteger. RFC 7748 uses little-endian; XECPublicKey.getU()
  // returns the numeric value. Valid u < 2^255, so no high-bit ambiguity.

  internal fun rawToPublicKey(raw: ByteArray): java.security.PublicKey {
    // little-endian -> BigInteger u (mask high bit per RFC 7748 on decode)
    val le = raw.copyOf(PUB_SIZE)
    le[PUB_SIZE - 1] = (le[PUB_SIZE - 1].toInt() and 0x7f).toByte()
    val be = ByteArray(PUB_SIZE)
    for (i in 0 until PUB_SIZE) be[i] = le[PUB_SIZE - 1 - i]
    val u = BigInteger(1, be)
    val spec = XECPublicKeySpec(NamedParameterSpec.X25519, u)
    return KeyFactory.getInstance("XDH").generatePublic(spec)
  }

  internal fun publicKeyToRaw(pub: XECPublicKey): ByteArray {
    val u = pub.u
    var be = u.toByteArray() // big-endian, may have sign byte / be shorter
    if (be.size > PUB_SIZE) be = be.copyOfRange(be.size - PUB_SIZE, be.size)
    val beFixed = ByteArray(PUB_SIZE)
    System.arraycopy(be, 0, beFixed, PUB_SIZE - be.size, be.size)
    val le = ByteArray(PUB_SIZE)
    for (i in 0 until PUB_SIZE) le[i] = beFixed[PUB_SIZE - 1 - i]
    return le
  }

  private fun decodeBase64Flexible(s: String): ByteArray {
    val t = s.trim()
    return try {
      if (t.contains('-') || t.contains('_')) {
        JavaBase64.getUrlDecoder().decode(t.padEnd((t.length + 3) / 4 * 4, '='))
      } else {
        JavaBase64.getDecoder().decode(t)
      }
    } catch (_: IllegalArgumentException) {
      // Last resort: url-decode after normalizing.
      JavaBase64.getUrlDecoder().decode(t.replace('+', '-').replace('/', '_'))
    }
  }
}
